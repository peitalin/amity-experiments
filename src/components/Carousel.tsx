
import * as React from 'react'
import 'styles/Carousel.scss'
import className from 'classname'

interface ReactProps {
  className?: string
}

export class Carousel extends React.Component<ReactProps, any> {

  state = {}

  render() {
    let className = this.props.className
      ? `${this.props.className} carousel__container`
      : 'carousel__container'
    return (
      <div className={className}>
        <div className="carousel__row">
          <div className="row__inner">
            { this.props.children }
          </div>
        </div>
      </div>
    )
  }
}

export default Carousel
