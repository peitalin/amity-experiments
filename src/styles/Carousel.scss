

$duration: 400ms;
$tileHeight: 240px;
$tileWidth: 160px;
$iconSize: 50px;
$growFactor: 1.1; // 1.3 times the original size

$moveLeft: -($tileWidth * ($growFactor - 1) / 2);
$moveRight: ($tileWidth * ($growFactor - 1) / 1);


$carousel-height: 300px;
$primary-color: #96616B;
// $primary-color: #1BD1C1;
// $slider-color: #F8BD7F;
$slider-color: #96616B;

// hide scrollbar
::-webkit-scrollbar {
  display: none;
}


.carousel__container {
  width: 100%;
  overflow: hidden;
  height: $carousel-height;
}

.carousel__row {
  overflow-x: scroll;
  overflow-y: hidden;
  width: 100%;
  height: $carousel-height;
}

.row__inner {
  font-size: 0;
  white-space: nowrap;
  margin: ($tileHeight / 8) 0;
  transition: transform 0.2s ease;
}

.tile__container {
  position: relative;
  display: inline-block;
  width: $tileWidth;
  height: $tileHeight;
  margin-right: 4px;
  font-size: 1rem;
  cursor: pointer;
  transition: transform $duration ease;
  transform-origin: center left;
}
.tile-right {
  transform-origin: center right;
}
.tile__img {
  width: $tileWidth;
  height: $tileHeight;
  object-fit: cover;
}

.tile__details {
  position: absolute;
  overflow: hidden;
  border: 2px solid $primary-color;
  padding: 1%;
  bottom: 0;
  left: 0;
  right: 0;
  top: 0;
  background: linear-gradient(to top, rgba(0,0,0,0.7) 0%, rgba(0,0,0,0) 100%);
  transition: $duration opacity;
}

.tile__title {
  position: absolute;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  width: 100%;
  bottom: 0;
  padding: 10px;
  font-size: 1rem;
  color: #eee;
  transition: transform 0.2s ease;
  transform-origin: left;
}
.tile__details:hover {
  .tile__title {
    transform: scale(0.8);
    transition: transform 0.2s ease;
  }
}

// *
// * Hover effects
// *

// We assume when you hover .row, you're also hovering a .tile
// First, move the whole row towards the left
.row__inner:hover {
  transform: translate3d($moveLeft, 0, 0);
  transition: transform 0.3s ease;
  transform-origin: center left;
}

.row__inner:hover .tile__container {
  // Fade out every tile
  opacity: .8;
  transition: transform 0.3s ease;
  // And then grow the current tile and make it opaque again
  &:hover {
    transform: scale($growFactor);
    transition: transform 0.3s ease;
    opacity: 1;
  }
}

// And finally move all tiles right of the current tile to the right,
// so its flush against the hovered tile
.tile__container:hover ~ .tile__container {
  transform: translate3d($moveRight, 0, 0);
}

