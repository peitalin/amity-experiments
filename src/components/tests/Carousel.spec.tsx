

import * as React from 'react'
import * as renderer from 'react-test-renderer'
import { mount, shallow } from 'enzyme'

import { Carousel } from '../Carousel'


class CarouselTest extends React.Component<ReactProps, any> {
  render() {
    return (
      <Carousel>
        <div style={{ height: 100, width: 100 }}></div>
        <div style={{ height: 100, width: 100 }}></div>
      </Carousel>
    )
  }
}

test('<Carousel/> component should contain 2 boxes', () => {
  const el = shallow( <CarouselTest/> )
  expect(el.find('div').length).toEqual(2)
})

test('<Carousel/> component matches snapshot', () => {
  const elR = renderer.create( <CarouselTest/> )
  expect(elR.toJSON()).toMatchSnapshot()
})

