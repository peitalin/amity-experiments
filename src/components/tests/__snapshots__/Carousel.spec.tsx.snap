// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel/> component matches snapshot 1`] = `
<div
  className="carousel__container"
>
  <div
    className="carousel__row"
  >
    <div
      className="row__inner"
    >
      <div
        style={
          Object {
            "height": 100,
            "width": 100,
          }
        }
      />
      <div
        style={
          Object {
            "height": 100,
            "width": 100,
          }
        }
      />
    </div>
  </div>
</div>
`;
